### PR TITLE
Folder restructuring and DDPG (+ Kipp, Tashi changes)

### PR DESCRIPTION
**Pull request by tombewley**

### CODE REVIEW

File: rlutils/agents/reinforce.py [hunk 1]
```diff
@@ -17,35 +19,38 @@ class ReinforceAgent:
...
         """Probabilistic action selection."""
+        state = state.to(self.device)
         if self.V is not None: action_probs, value = self.pi(state), self.V(state)
         else: action_probs = self.pi(state)
         dist = Categorical(action_probs) # Categorical action distribution.
         action = dist.sample()
-        if self.V is not None: self.ep_predictions.append((dist.log_prob(action), value[0]))
-        else: self.ep_predictions.append(dist.log_prob(action))
-        return action
+        extra = {"pi": action_probs.detach().numpy()}
```
Comment (kippfreud) [inline rlutils/agents/reinforce.py]: Does not work when using cuda - need to move to CPU from GPU first. Use 
```
extra = {"pi": action_probs.cpu().detach().numpy()}
```

File: rlutils/agents/actor_critic.py [hunk 2]
```diff
@@ -16,32 +18,29 @@ class ActorCriticAgent:
...
 
-    def act(self, state):
+    def act(self, state, explore=True):
         """Probabilistic action selection."""
         action_probs, value = self.pi(state), self.V(state)
         dist = Categorical(action_probs) # Categorical action distribution.
         action = dist.sample()
         self.last_s_l_v = (state, dist.log_prob(action), value[0])
-        return action
+        return action, {"pi": action_probs.detach().numpy(), "V": value[0].item()}
```
Comment (kippfreud) [inline rlutils/agents/actor_critic.py]: should be
```
action_probs.cpu().detach().numpy()
```
cos this doesn't work if action probs is on GPU

File: rlutils/agents/actor_critic.py [hunk 3]
```diff
@@ -16,32 +18,29 @@ class ActorCriticAgent:
...
+        self.V = SequentialNetwork(preset=preset_V, input_shape=state_shape, output_size=1).to(self.device)
         self.optimiser_V = optim.Adam(self.V.parameters(), lr=self.P["lr_V"])
-        self.eps = np.finfo(np.float32).eps.item() # Small float used to prevent div/0 errors.
         # Tracking variables.
         self.last_s_l_v = None # State, log prob action and value.
         self.ep_losses = []
 
-    def act(self, state):
+    def act(self, state, explore=True):
         """Probabilistic action selection."""
```
Comment (kippfreud) [inline rlutils/agents/actor_critic.py]: first add 
```
state = state.to(self.device)
```
cos this doesn't work with GPU

File: rlutils/agents/actor_critic.py [hunk 4]
```diff
@@ -16,32 +18,29 @@ class ActorCriticAgent:
...
         action_probs, value = self.pi(state), self.V(state)
         dist = Categorical(action_probs) # Categorical action distribution.
         action = dist.sample()
         self.last_s_l_v = (state, dist.log_prob(action), value[0])
-        return action
+        return action, {"pi": action_probs.detach().numpy(), "V": value[0].item()}
 
     def update_on_transition(self, next_state, reward):
```
Comment (kippfreud) [inline rlutils/agents/actor_critic.py]: next state needs to be moved to device:
```
next_state = next_state.to(self.device)
```

Comment (kippfreud) [reply rlutils/agents/actor_critic.py]: actually, only perform above code if next state is not none

File: rlutils/agents/ddpg.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class DdpgAgent:
+    def __init__(self, 
+                 state_shape,
+                 action_space, 
+                 hyperparameters=DEFAULT_HYPERPARAMETERS
+                 ):
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self.P = hyperparameters
```
Comment (kippfreud) [inline rlutils/agents/ddpg.py]: Might be worth commenting these variables? I don't know what the internal params are - could be labelled. this is true for all agent init functions.

File: BC4.txt [hunk 6]
```diff
@@ -0,0 +1,116 @@
+###################To get access
```
Comment (tashinam) [inline BC4.txt]: I don't think you should put this in public domain. Perhaps put this in .gitignore

**Pull request merged**